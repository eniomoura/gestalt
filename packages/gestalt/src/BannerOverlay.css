.textColorOverrideLight * {
  /* stylelint-disable-next-line declaration-no-important */
  color: var(--color-text-dark) !important;
}

.textErrorColorOverrideLight * {
  /* stylelint-disable-next-line declaration-no-important */
  color: var(--color-text-dark) !important;
  /* stylelint-disable-next-line declaration-no-important */
  font-weight: var(--font-weight-semibold) !important;
}

.textColorOverrideDark * {
  /* stylelint-disable-next-line declaration-no-important */
  color: var(--color-text-dark) !important;
}

.parentButton {
  composes: noBorder from "./Borders.css";
  composes: p0 from "./Whitespace.css";
  composes: paddingX0 from "./boxWhitespace.css";
  composes: paddingY0 from "./boxWhitespace.css";
  composes: transparentBg from "./Colors.css";
}

