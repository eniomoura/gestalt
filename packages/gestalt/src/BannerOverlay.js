// @flow strict
import { Children, type Element, type ElementConfig, type Node as ReactNode } from 'react';
import classnames from 'classnames';
import Avatar from './Avatar';
import styles from './BannerOverlay.css';
import PrimaryAction from './BannerOverlay/PrimaryAction';
import {
  BannerOverlayAvatarThumbnail,
  BannerOverlayIconThumbnail,
  BannerOverlayImageThumbnail,
  BannerOverlayMessage,
  BannerOverlayTypeThumbnail,
} from './BannerOverlay/subcomponents';
import Box from './Box';
import Button from './Button';
import ButtonLink from './ButtonLink';
import { useColorScheme } from './contexts/ColorSchemeProvider';
import Flex from './Flex';
import Icon from './Icon';
import Image from './Image';
import Layer from './Layer';
import Link from './Link';
import Text from './Text';
import useResponsiveMinWidth from './useResponsiveMinWidth';

const DEFAULT_COLORS = {
  containerColor: 'default',
  textColor: 'default',
  iconColor: 'white',
};

const COLORS_BY_TYPE = Object.freeze({
  default: DEFAULT_COLORS,
  success: DEFAULT_COLORS,
  error: { ...DEFAULT_COLORS, containerColor: 'errorBase' },
  progress: {
    containerColor: 'secondary',
    textColor: 'default',
    iconColor: 'darkGray',
  },
});

type Props = {
  /**
   * Main content of BannerOverlay. Content should be [localized](https://gestalt.pinterest.systems/web/banneroverlay#Localization). See the [Text variant](https://gestalt.pinterest.systems/web/banneroverlay#Text) to learn more.
   */
  message: string | Element<typeof Text>,
  /**
   * Secndary content of BannerOverlay. Content should be [localized](https://gestalt.pinterest.systems/web/banneroverlay#Localization). See the [Text variant](https://gestalt.pinterest.systems/web/banneroverlay#Text) to learn more.
   */
  title: string | Element<typeof Text>,
  /**
   * Adds a dismiss button to BannerOverlay. See the [Dismissible variant](https://gestalt.pinterest.systems/web/banneroverlay#Dismissible) for more info.
   * The `accessibilityLabel` should follow the [Accessibility guidelines](https://gestalt.pinterest.systems/web/banneroverlay#Accessibility).
   *
   */
  dismissButton?: {
    accessibilityLabel?: string,
    onDismiss: () => void,
  },
  /**
   * Helper [Link](https://gestalt.pinterest.systems/web/link) to be placed after the subtext. See the [helper link variant](https://gestalt.pinterest.systems/web/banneroverlay#helperLink) to learn more.
   */
  helperLink?: {
    text: string,
    accessibilityLabel: string,
    href: string,
    onClick?: ({
      event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement>,
      dangerouslyDisableOnNavigation: () => void,
    }) => void,
  },
  /**
   * Adds an optional button for user interaction. Generally not recommended given the ephemeral nature of BannerOverlays.
   */
  primaryAction?:
    | {
        accessibilityLabel: string,
        href: string,
        label: string,
        onClick?: $ElementType<ElementConfig<typeof ButtonLink>, 'onClick'>,
        rel?: $ElementType<ElementConfig<typeof Link>, 'rel'>,
        role: 'link',
        size?: $ElementType<ElementConfig<typeof Button>, 'size'>,
        target?: $ElementType<ElementConfig<typeof Link>, 'target'>,
      }
    | {
        accessibilityLabel: string,
        label: string,
        onClick: $ElementType<ElementConfig<typeof Button>, 'onClick'>,
        role?: 'button',
        size?: $ElementType<ElementConfig<typeof Button>, 'size'>,
      },

  /**
   * An optional thumbnail to display next to the text.
   */
  thumbnail?:
    | { image: Element<typeof Image> }
    | { avatar: Element<typeof Avatar> }
    | { icon: Element<typeof Icon> },
  /**
   * See the [type variant](https://gestalt.pinterest.systems/web/banneroverlay#Type) to learn more.
   */
  type?: 'default' | 'success' | 'error' | 'progress',
};

/**
 * [BannerOverlays](https://gestalt.pinterest.systems/web/banneroverlay) displays short educational messages when users have performed actions that indicate some intent, such as a related pin tap or idea pin swipe. It is scrim-less, meaning users can scroll content underneath without having a wash layer on top of the content.
 *
 * BannerOverlay is a sticky component triggered by scroll, on tap or long-press, placed at the bottom of the screen.
 * ![BannerOverlay light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/BannerOverlay.spec.mjs-snapshots/BannerOverlay-chromium-darwin.png)
 */
export default function BannerOverlay({
  message,
  title,
  dismissButton,
  helperLink,
  primaryAction,
  thumbnail,
  type = 'default',
}: Props): ReactNode {
  const { name: colorSchemeName } = useColorScheme();
  const isDarkMode = colorSchemeName === 'darkMode';

  const responsiveMinWidth = useResponsiveMinWidth();
  const isMobileWidth = responsiveMinWidth === 'xs';

  let messageTextElement: Element<'span'> | string;

  if (typeof message === 'string') {
    messageTextElement = message;
  }
  if (typeof title === 'string') {
    messageTextElement = title;
  }

  // If `message` is a Text component, we need to override any text colors within to ensure they all match
  const messageIsTextNode =
    typeof message !== 'string' &&
    Children.only<Element<typeof Text>>(message).type.displayName === 'Text';

  if (messageIsTextNode) {
    const textColorOverrideStyles = isDarkMode
      ? styles.textColorOverrideDark
      : styles.textColorOverrideLight;

    messageTextElement = <span className={textColorOverrideStyles}>{message}</span>;
  }

  // TODO: Add BannerOverlay to DefautlLabelProvider
  // const { accessibilityDismissButtonLabel: accessibilityDismissButtonLabelDefault } =
  //   useDefaultLabelContext('BannerOverlay');
  const accessibilityDismissButtonLabelDefault = '';

  const { containerColor, textColor } = COLORS_BY_TYPE[type];

  const isDefaultBannerOverlay = type === 'default';
  const isNotDefaultBannerOverlay = ['success', 'error', 'progress'].includes(type);

  return (
    <Layer>
      <Box
        color={containerColor}
        paddingX={4}
        paddingY={3}
        rounding={4}
        dangerouslySetInlineStyle={{
          __style: {
            border: '1px solid var(--color-gray-roboflow-200)',
            bottom: 50,
            left: '50%',
            transform: 'translateX(-50%)',
          },
        }}
        position="fixed"
        display="flex"
        justifyContent="center"
        direction="column"
        fit
        minWidth={isMobileWidth ? 348 : 900}
      >
        {dismissButton ? (
          <Flex.Item flex="none" alignSelf="end">
            <button
              aria-label={
                dismissButton.accessibilityLabel ?? accessibilityDismissButtonLabelDefault
              }
              className={classnames(styles.parentButton)}
              onClick={dismissButton.onDismiss}
              type="button"
            >
              X
            </button>
          </Flex.Item>
        ) : null}
        <Flex alignItems="center" gap={4}>
          {isDefaultBannerOverlay &&
          !!thumbnail?.image &&
          Children.only<Element<typeof Image>>(thumbnail.image).type.displayName === 'Image' ? (
            <Flex.Item flex="none">
              <BannerOverlayImageThumbnail thumbnail={thumbnail.image} />
            </Flex.Item>
          ) : null}

          {isDefaultBannerOverlay &&
          !!thumbnail?.icon &&
          Children.only<Element<typeof Icon>>(thumbnail.icon).type.displayName === 'Icon' ? (
            <Flex.Item flex="none">
              <BannerOverlayIconThumbnail thumbnail={thumbnail.icon} />
            </Flex.Item>
          ) : null}

          {isDefaultBannerOverlay &&
          !!thumbnail?.avatar &&
          Children.only<Element<typeof Avatar>>(thumbnail.avatar).type.displayName === 'Avatar' ? (
            <Flex.Item flex="none">
              <BannerOverlayAvatarThumbnail thumbnail={thumbnail.avatar} />
            </Flex.Item>
          ) : null}

          {isNotDefaultBannerOverlay ? (
            <Flex.Item flex="none">
              <BannerOverlayTypeThumbnail type={type} />
            </Flex.Item>
          ) : null}

          <Flex.Item flex="grow">
            <Text weight="bold">{title}</Text>
            <BannerOverlayMessage
              text={messageIsTextNode ? undefined : messageTextElement}
              textElement={messageIsTextNode ? messageTextElement : undefined}
              helperLink={helperLink}
              textColor={textColor}
              type={type}
            />
          </Flex.Item>
        </Flex>
        {primaryAction ? (
          // Allow button text to wrap on mobile
          <Flex.Item flex={isMobileWidth ? 'shrink' : 'none'} alignSelf="end">
            {primaryAction?.accessibilityLabel && primaryAction?.label
              ? (primaryAction.role === 'link' && (
                  <PrimaryAction
                    accessibilityLabel={primaryAction.accessibilityLabel}
                    href={primaryAction.href}
                    label={primaryAction.label}
                    onClick={primaryAction.onClick}
                    rel={primaryAction?.rel}
                    role="link"
                    size="sm"
                    target={primaryAction?.target}
                  />
                ),
                primaryAction.role !== 'link' && (
                  <PrimaryAction
                    accessibilityLabel={primaryAction.accessibilityLabel}
                    label={primaryAction.label}
                    onClick={primaryAction.onClick}
                    role="button"
                    size="sm"
                  />
                ))
              : null}
          </Flex.Item>
        ) : null}
      </Box>
    </Layer>
  );
}
