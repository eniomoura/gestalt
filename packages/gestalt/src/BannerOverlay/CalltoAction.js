// @flow strict
import { type ElementConfig, type Node as ReactNode } from 'react';
import Button from '../Button';
import ButtonLink from '../ButtonLink';

type Props =
  | {
      accessibilityLabel: string,
      color: 'red' | 'gray',
      href: string,
      label: string,
      onClick?: $ElementType<ElementConfig<typeof ButtonLink>, 'onClick'>,
      rel?: $ElementType<ElementConfig<typeof ButtonLink>, 'rel'>,
      role: 'link',
      size?: $ElementType<ElementConfig<typeof ButtonLink>, 'size'>,
      target?: $ElementType<ElementConfig<typeof ButtonLink>, 'target'>,
    }
  | {
      accessibilityLabel: string,
      color: 'red' | 'gray',
      label: string,
      onClick: $ElementType<ElementConfig<typeof Button>, 'onClick'>,
      role: 'button',
      size?: $ElementType<ElementConfig<typeof Button>, 'size'>,
    };

export default function CallToAction({
  accessibilityLabel,
  label,
  size = 'lg',
  ...props
}: Props): ReactNode {
  if (props.role === 'link')
    return (
      <ButtonLink
        accessibilityLabel={accessibilityLabel}
        color={props.color}
        href={props.href ?? ''}
        onClick={props.onClick}
        rel={props.rel}
        target={props.target}
        text={label}
        size={size}
      />
    );

  return (
    <Button
      accessibilityLabel={accessibilityLabel}
      color={props.color}
      onClick={props.onClick}
      size={size}
      text={label}
    />
  );
}
