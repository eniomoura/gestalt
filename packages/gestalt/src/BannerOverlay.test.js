// @flow strict
import ReactDOM from 'react-dom';
import { create } from 'react-test-renderer';
import BannerOverlay from './BannerOverlay';
import Image from './Image';
import Link from './Link';
import Text from './Text';

describe('<BannerOverlay />', () => {
  test('Text Only', () => {
    // (ReactDOM.createPortal: JestMockFn<[Element], Element).mockImplementation(
    //   (e) => e,
    // );

    const tree = create(
      <BannerOverlay
        title="Profile"
        message="Same great profile, slightly new look. Learn more?"
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Text + Image', () => {
    const tree = create(
      <BannerOverlay
        title="Text and Image"
        thumbnail={{
          image: (
            <Image
              alt=""
              naturalHeight={1}
              naturalWidth={1}
              src="https://i.pinimg.com/474x/b2/55/ed/b255edbf773ffb3985394e6efb9d2a49.jpg"
            />
          ),
        }}
        message={
          <Text inline weight="bold">
            Saved to{' '}
            <Link
              display="inlineBlock"
              href="https://www.pinterest.com/search/pins/?q=home%20decor"
              underline="hover"
            >
              Home decor
            </Link>
          </Text>
        }
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Text + Image + Button', () => {
    const tree = create(
      <BannerOverlay
        title="Text, image and button"
        thumbnail={{
          image: (
            <Image
              alt=""
              naturalHeight={1}
              naturalWidth={1}
              src="https://i.pinimg.com/474x/b2/55/ed/b255edbf773ffb3985394e6efb9d2a49.jpg"
            />
          ),
        }}
        message={
          <Text inline weight="bold">
            Saved to{' '}
            <Link
              display="inlineBlock"
              href="https://www.pinterest.com/search/pins/?q=home%20decor"
              underline="hover"
            >
              Home decor
            </Link>
          </Text>
        }
        primaryAction={{
          accessibilityLabel: 'Undo undo action',
          label: 'Undo',
          size: 'lg',
          role: 'button',
          onClick: () => {},
        }}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
