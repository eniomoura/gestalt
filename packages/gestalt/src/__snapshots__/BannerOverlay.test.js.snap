// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BannerOverlay /> Text + Image + Button 1`] = `<div />`;

exports[`<BannerOverlay /> Text + Image 1`] = `<div />`;

exports[`<BannerOverlay /> Text Only 1`] = `<div />`;
